File: java/0006.Z字变换.java
class Solution {
    // 1. O(n) t:2ms(99%) O(1) m:41.6M(51%)
    public String convert(String s, int numRows) {
        int m = numRows * 2 - 2, n = s.length();
        if (n == 1 || numRows == 1) return s;
        int idx = 0;
        StringBuilder ans = new StringBuilder();
        while (idx < n){
            ans.append(s.charAt(idx));
            idx += m;
        }
        for (int i = 1; i < numRows - 1; i++){
            idx = i;
            while (idx < n){
                ans.append(s.charAt(idx));
                if (idx - i + m - i < n)
                    ans.append(s.charAt(idx - i + m - i));
                idx += m;
            }
        }
        idx = numRows - 1;
        while (idx < n){
            ans.append(s.charAt(idx));
            idx += m;
        }
        return ans.toString();
    }
}

File: python/0006.Z字变换.py
class Solution:
    # 1.O(n) t:52ms(80%) O(1) m:15M(87%) 模拟
    def convert(self, s: str, numRows: int) -> str:
        m, n = numRows * 2 - 2, len(s)
        if n == 1 or numRows == 1: return s
        ans, idx = '', 0
        while idx < n:
            ans += s[idx]
            idx += m
        for i in range(1, numRows - 1):
            idx = i
            while idx < n:
                ans += s[idx]
                if idx - i + m - i < n:
                    ans += s[idx - i + m - i]
                idx += m
        idx = numRows - 1
        while idx < n:
            ans += s[idx]
            idx += m
        return ans